solarforecastarbiter/metrics/deterministic.py

```diff
@@ -545,6 +549,151 @@ def combined_performance_index(obs, fx):
+    ).set_index('timestamp')['cost'].tz_localize(tz).sort_index()
+    # only get those values at index filling as appropriate
+    ser = base_ser.reindex(index, method=fill)
+    return ser
+
+
+def _time_of_day_cost(obs, fx, cost_params, error_fnc):
+    """Compute cost according to a datamodel.TimeOfDayCost"""
+    agg_fnc = _np_agg_fnc(cost_params.aggregation, cost_params.net)
+    fill = _FILL_OPTIONS[cost_params.fill]
+
+    errors = error_fnc(obs, fx)
+    tz = cost_params.timezone or errors.index.tzinfo
+    cost_ser = _make_time_of_day_cost_ser(
+        cost_params.times, cost_params.costs, errors.index, tz, fill)
+    error_cost = errors * cost_ser
+    return agg_fnc(error_cost)
+
+
+def _datetime_cost(obs, fx, cost_params, error_fnc):
+    """Compute cost according to a datamodel.DatetimeCost"""
+    agg_fnc = _np_agg_fnc(cost_params.aggregation, cost_params.net)
+    fill = _FILL_OPTIONS[cost_params.fill]
+    cost_ser = pd.Series(cost_params.costs,
+                         index=pd.DatetimeIndex(cost_params.datetimes),
+                         dtype=float)
+
+    errors = error_fnc(obs, fx)
+
+    if cost_ser.index.tzinfo is None:
+        if cost_params.timezone is not None:
+            cost_ser = cost_ser.tz_localize(cost_params.timezone)
+        else:
+            cost_ser = cost_ser.tz_localize(errors.index.tzinfo)
+
+    error_cost = errors * cost_ser.reindex(errors.index, method=fill)
+    return agg_fnc(error_cost)
+
+
+def _band_masks(bands, errors):
+    """Make masks for each band based on which band errors falls in"""
+    prev = np.zeros(errors.shape, dtype=bool)
+    out = []
+    for band in bands:
+        emin, emax = band.error_range
+        new = (errors >= emin) & (errors <= emax)
+        # only those new locations that not also in prev should be used
+        both = prev & new
+        new[both] = False
+        out.append(new)
+        prev |= new
+    return out
+
+
+def _error_band_cost(obs, fx, cost_params, error_fnc):
+    """Calculate cost using datamodel.BandedCost parameters"""
+    bands = cost_params.bands
+    band_cost_functions = [
+        partial(_COST_FUNCTION_MAP[band.cost_function],
+                cost_params=band.cost_function_parameters,
+                error_fnc=error_fnc)
+        for band in bands
+    ]
+
+    errors = error_fnc(obs, fx)
+    out = 0
+    masks = _band_masks(bands, errors)
+    for mask, fnc in zip(masks, band_cost_functions):
+        if not mask.any():
+            continue
+        mobs = obs[mask]
+        mfx = fx[mask]
+        out += fnc(mobs, mfx)
+    return out
+
+
+def cost(obs, fx, cost_params, error_fnc=error):
+    """
+    GOODER DOCS
+    """
+    fnc = _COST_FUNCTION_MAP[cost_params.type]
+    return fnc(obs, fx, cost_params.parameters, error_fnc)
+
+
+_COST_FUNCTION_MAP = {
+    'constant': _constant_cost,
+    'timeofday': _time_of_day_cost,
+    'datetime': _datetime_cost,
+    'errorband': _error_band_cost
```

Comment:
I'm ok with private functions, but seems like these could be useful public functions too.

Reply:
sure